ftwhttp/header.go

```diff
@@ -4,149 +4,182 @@
 package ftwhttp
 
 import (
-	"bytes"
+	"bufio"
 	"io"
 	"net/textproto"
-	"sort"
+	"slices"
+	"strings"
 
 	"github.com/rs/zerolog/log"
 )
 
 const (
-	// ContentTypeHeader gives you the string for content type
-	ContentTypeHeader string = "Content-Type"
+	headerSeparator = ": "
+	headerDelimiter = "\r\n"
 )
 
-// Based on https://golang.org/src/net/http/header.go
-
-// Header is a simplified version of headers, where there is only one header per key.
-// The original golang stdlib uses a proper string slice to map this.
-type Header map[string]string
-
-// stringWriter implements WriteString on a Writer.
-type stringWriter struct {
-	w io.Writer
+// Header is a representation of the HTTP header section.
+// It holds an ordered list of HeaderTuples.
+type Header struct {
+	canonicalNames map[string]uint
+	entries        []HeaderTuple
 }
 
-// WriteString writes the string on a Writer
-func (w stringWriter) WriteString(s string) (n int, err error) {
-	return w.w.Write([]byte(s))
+// HeaderTuple is a representation of an HTTP header. It consists
+// of a name and value.
+type HeaderTuple struct {
+	Name  string
+	Value string
 }
 
-// Add adds the (key, value) pair to the headers if it does not exist
-// The key is case-insensitive
-func (h Header) Add(key, value string) {
-	if h.Get(key) == "" {
-		h.Set(key, value)
+// Creates an empty Header. You should not initialize the struct directly.
+func NewHeader() *Header {
+	return &Header{
+		canonicalNames: map[string]uint{},
+		entries:        []HeaderTuple{},
 	}
 }
 
-// Set sets the header entries associated with key to
-// the single element value. It replaces any existing
-// values associated with a case-insensitive key.
-func (h Header) Set(key, value string) {
-	h.Del(key)
-	h[key] = value
-}
-
-// Get gets the value associated with the given key.
-// If there are no values associated with the key, Get returns "".
-// The key is case-insensitive
-func (h Header) Get(key string) string {
-	if h == nil {
-		return ""
+// Creates a new Header from a map of HTTP header names and values.
+//
+// This is a convenience and legacy fallback method. In the future,
+// headers should be specified as a list, in order to guarantee order
+// and to allow requests to contain the same header multiple times,
+// potentially, but not necessarily, with different values.
```

Comment:
```suggestion
// potentially with different values.
```
I think that _potentially_ already offers the needed conditional future.